File: go-spider/internal/security/redirect_policy.go
package security

import (
	"net/url"
	"strings"
	"unicode"
)

// RedirectPolicyError is a stable redirect planning failure.
type RedirectPolicyError struct {
	ReasonCode string
}

func (e *RedirectPolicyError) Error() string { return e.ReasonCode }

func redirectFail(reason string) error { return &RedirectPolicyError{ReasonCode: reason} }

// RedirectHop is one fully revalidated redirect hop.
type RedirectHop struct {
	Hop       int
	URL       string
	Decision  PolicyDecision
	Target    PinnedTarget
	Addresses []string
}

// RedirectPlan is an ordered list of planned redirect hops.
type RedirectPlan struct {
	Hops []RedirectHop
}

var redirectSyntaxReasons = map[string]bool{
	"absolute_url_required":    true,
	"ambiguous_ip_literal":     true,
	"authority_required":       true,
	"backslash_not_allowed":    true,
	"control_character":        true,
	"empty_url":                true,
	"host_too_long":            true,
	"invalid_host":             true,
	"invalid_idna":             true,
	"invalid_percent_encoding": true,
	"invalid_port":             true,
	"invalid_scheme":           true,
	"invalid_url":              true,
	"label_too_long":           true,
	"scheme_not_allowed":       true,
	"url_too_long":             true,
}

func normalizeRedirectReason(reason string) string {
	if redirectSyntaxReasons[reason] {
		return "redirect_location_invalid"
	}
	return reason
}

func containsControlOrFormat(value string) bool {
	for _, r := range value {
		if unicode.IsControl(r) || unicode.Is(unicode.Cf, r) {
			return true
		}
	}
	return false
}

// ResolveRedirectLocation resolves relative Locations against the current URL.
func ResolveRedirectLocation(currentURL, location string) (string, error) {
	if location == "" || strings.TrimSpace(location) == "" {
		return "", redirectFail("redirect_location_missing")
	}
	if location != strings.TrimSpace(location) {
		return "", redirectFail("redirect_location_invalid")
	}
	if containsControlOrFormat(location) {
		return "", redirectFail("redirect_location_invalid")
	}
	base, err := url.Parse(currentURL)
	if err != nil {
		return "", redirectFail("redirect_location_invalid")
	}
	ref, err := url.Parse(location)
	if err != nil {
		return "", redirectFail("redirect_location_invalid")
	}
	resolved := base.ResolveReference(ref)
	if !resolved.IsAbs() {
		return "", redirectFail("redirect_location_invalid")
	}
	return resolved.String(), nil
}

// ResolveRedirectLocationValue validates JSON-style Location values at the
// shared fixture boundary while keeping PlanRedirectHop strongly typed.
func ResolveRedirectLocationValue(currentURL string, location any) (string, error) {
	if location == nil {
		return "", redirectFail("redirect_location_missing")
	}
	text, ok := location.(string)
	if !ok {
		return "", redirectFail("redirect_location_invalid")
	}
	return ResolveRedirectLocation(currentURL, text)
}

// PlanRedirectHop revalidates one redirect hop with a fresh DNS/policy decision
// and a fresh PinnedTarget.
func PlanRedirectHop(currentURL, location string, policy OutboundPolicy, resolver Resolver, hop int) (RedirectHop, error) {
	if hop <= 0 {
		return RedirectHop{}, redirectFail("invalid_redirect_hop")
	}
	if hop > MaxRedirects {
		return RedirectHop{}, redirectFail("redirect_limit_exceeded")
	}
	targetRaw, err := ResolveRedirectLocation(currentURL, location)
	if err != nil {
		return RedirectHop{}, err
	}
	decision, err := DecideRedirect(currentURL, targetRaw, policy, resolver)
	if err != nil {
		return RedirectHop{}, redirectFail("redirect_location_invalid")
	}
	if !decision.Allowed {
		return RedirectHop{}, redirectFail(normalizeRedirectReason(decision.ReasonCode))
	}
	if decision.Normalized == nil {
		return RedirectHop{}, redirectFail("redirect_location_invalid")
	}
	target, err := NewPinnedTarget(decision, policy)
	if err != nil {
		return RedirectHop{}, err
	}
	return RedirectHop{
		Hop:       hop,
		URL:       decision.Normalized.NormalizedURL,
		Decision:  decision,
		Target:    target,
		Addresses: append([]string(nil), decision.Addresses...),
	}, nil
}

// PlanRedirects revalidates a sequence of redirect Locations.
func PlanRedirects(currentURL string, locations []string, policy OutboundPolicy, resolver Resolver) (RedirectPlan, error) {
	hops := make([]RedirectHop, 0, len(locations))
	current := currentURL
	for index, location := range locations {
		hopNumber := index + 1
		if hopNumber > MaxRedirects {
			return RedirectPlan{}, redirectFail("redirect_limit_exceeded")
		}
		hop, err := PlanRedirectHop(current, location, policy, resolver, hopNumber)
		if err != nil {
			return RedirectPlan{}, err
		}
		hops = append(hops, hop)
		current = hop.URL
	}
	return RedirectPlan{Hops: hops}, nil
}

File: crawler/security/redirect_policy.py
"""Redirect per-hop revalidation planner for TASK-022D-1.

The planner is pure decision logic: it resolves each Location, re-runs the
TASK-022B/C policy pipeline, and produces a fresh PinnedTarget for every hop.
It never performs network requests, DNS, redirect following, or body I/O.
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass
from typing import Sequence
from urllib.parse import urljoin, urlsplit

from crawler.security.models import NormalizedURL, OutboundPolicy, PolicyDecision
from crawler.security.outbound_policy import decide
from crawler.security.pinned_connection import PinnedTarget, build_pinned_target
from crawler.security.transport_budget import MAX_REDIRECTS

_LOCATION_SYNTAX_REASONS = frozenset(
    {
        "absolute_url_required",
        "ambiguous_ip_literal",
        "authority_required",
        "backslash_not_allowed",
        "control_character",
        "empty_url",
        "host_too_long",
        "invalid_host",
        "invalid_idna",
        "invalid_percent_encoding",
        "invalid_port",
        "invalid_scheme",
        "invalid_url",
        "label_too_long",
        "scheme_not_allowed",
        "url_too_long",
    }
)


class RedirectPolicyError(ValueError):
    def __init__(self, reason_code: str, message: str = "") -> None:
        super().__init__(message or reason_code)
        self.reason_code = reason_code


@dataclass(frozen=True, slots=True)
class RedirectHop:
    hop: int
    url: str
    normalized: NormalizedURL
    decision: PolicyDecision
    pinned_target: PinnedTarget
    addresses: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class RedirectPlan:
    hops: tuple[RedirectHop, ...]
    final_hop: RedirectHop | None


def _resolve_location(current_url: str, location: str) -> str:
    if location is None:
        raise RedirectPolicyError("redirect_location_missing")
    if not isinstance(location, str):
        raise RedirectPolicyError("redirect_location_invalid")
    if location == "":
        raise RedirectPolicyError("redirect_location_missing")
    if location.strip() == "":
        raise RedirectPolicyError("redirect_location_missing")
    if location != location.strip():
        raise RedirectPolicyError("redirect_location_invalid")
    if any(unicodedata.category(char) in ("Cc", "Cf") for char in location):
        raise RedirectPolicyError("redirect_location_invalid")
    try:
        if location.startswith("//"):
            current_scheme = urlsplit(current_url).scheme
            if current_scheme not in ("http", "https"):
                raise RedirectPolicyError("redirect_location_invalid")
            return f"{current_scheme}:{location}"
        if "://" in location:
            return location
        return urljoin(current_url, location)
    except RedirectPolicyError:
        raise
    except Exception:
        raise RedirectPolicyError("redirect_location_invalid")


def plan_redirect_hop(
    current_url: str,
    location: str,
    policy: OutboundPolicy,
    *,
    resolver,
    hop_count: int = 1,
) -> RedirectHop:
    if isinstance(hop_count, bool) or not isinstance(hop_count, int) or hop_count <= 0:
        raise RedirectPolicyError("invalid_redirect_hop")
    if hop_count > MAX_REDIRECTS:
        raise RedirectPolicyError("redirect_limit_exceeded")

    target_raw = _resolve_location(current_url, location)
    decision = decide(target_raw, policy, resolver=resolver, previous_raw=current_url)
    if not decision.allowed:
        reason = decision.reason_code
        if reason in _LOCATION_SYNTAX_REASONS:
            raise RedirectPolicyError("redirect_location_invalid")
        raise RedirectPolicyError(reason)

    normalized = decision.normalized
    assert normalized is not None
    pinned_target = build_pinned_target(decision, policy, policy_identity="redirect")
    return RedirectHop(
        hop=hop_count,
        url=normalized.normalized_url,
        normalized=normalized,
        decision=decision,
        pinned_target=pinned_target,
        addresses=tuple(decision.addresses),
    )


def plan_redirects(
    current_url: str,
    locations: Sequence[str],
    policy: OutboundPolicy,
    *,
    resolver,
) -> RedirectPlan:
    hops: list[RedirectHop] = []
    current = current_url
    for index, location in enumerate(locations, start=1):
        if index > MAX_REDIRECTS:
            raise RedirectPolicyError("redirect_limit_exceeded")
        hop = plan_redirect_hop(current, location, policy, resolver=resolver, hop_count=index)
        hops.append(hop)
        current = hop.url
    return RedirectPlan(tuple(hops), hops[-1] if hops else None)
